runtime/lua/vim/filetype.lua

```diff
@@ -1043,6 +1132,14 @@ local extension = {
   sqi = 'sqr',
   sqr = 'sqr',
   nut = 'squirrel',
+  src = detect.src,
+  srC = detect.src,
+  sRc = detect.src,
+  sRC = detect.src,
+  Src = detect.src,
+  SrC = detect.src,
+  SRc = detect.src,
+  SRC = detect.src,
```

Comment:
This seems pretty onerous, and reduces readability. We're doing this for a 5% perf improvement?

Reply:
> This seems pretty onerous, and reduces readability. We're doing this for a 5% perf improvement?

5% is not *that* small improvement. Plus:

> There are at least two reasons for doing this:
> 
> - Performance. It will reduce the number of needed pattern matches (especially moving extensions, as those don't have a "fast" parent pattern). Profiling showed about 5% improvement on cases used in #29660.
> - My future plan to suggest adding `fast` option flag to `vim.filetype.match()` which will do only explicit filename and extension matching. This will allow about 10x speed improvement for plugins/users at the cost of not doing pattern and content matching. Having as much explicit filename and extension matches as possible will allow more filetype coverage.

I had to reschedule a bit my work on `vim.filetype.match()`. I have plans for fixing inconsistency in how `contents` is treated compared to `buf` and adding described `fast` option.

Reply:
The 10x improvement from "fast" doesn't require any of this manual expansion, does it?

Reply:
> The 10x improvement from `fast` doesn't require any of this manual pattern expansion, does it?

If left as pattern matching, they would not match with `fast` (as it is planned to just skip this phase). With explicit extension match (as after this PR) all those cases ("src", "SRC", etc.) will properly match even with `fast`.

Reply:
I would guess that 80% of the gains from `fast` is that it skips *content* matching. Pattern-matching names, especially simple things like case insensitivity, are unlikely to have a big perf cost. Right?

Reply:
> I would guess that 80% of the gains from `fast` is that it skips _content_ inspection. Pattern-matching names, especially simple things like case insensitivity, are unlikely to have a big perf cost. Right?

My benchmarking shows otherwise. See #29660.

Reply:
If the caller wants fast, then they don't need to match `SRc` and related variants. We shouldn't have to hardcode case permutations into our list of filenames.

Reply:
Besides effect on the (already suggested in #30086) `fast` option for `vim.filetype.match()`, reducing the number of times `vim.filetype.match()` does string pattern matching improves performance in the default `fast = false` case. As was already said, by about 5%, which is not quite negligible.

I indeed hesitated to suggest this change, but the current footprint of (12 pattern into 99 extensions) seemed borderline acceptable for the benefits it provides.

Also, adding new case-insensitive extension-like matches is *very* uncommon. All such changes seem to have been done in two commits in April 2022:
- https://github.com/vim/vim/commit/3ad2090316edc85e93094bba7af64f9991cc7f85
- https://github.com/vim/vim/commit/0bbf09ca41382302493e5db51b01d2fbdc778586

With this frequency, I don't think will add too much maintenance burden.